### ISSUE TITLE
plugin/rewrite: fallback names

### ISSUE THREAD

**Issue by achille-roussel** [coredns/coredns#1394]
Hello,

We're looking into coredns as a potential corner stone to support availability zone affinity on AWS. Our idea is to have rewrite rules that would translate queries for `<service>.<domain>` into `<zone>.<service>.service.consul`, with the zone being set at boot time from the EC2 instance metadata.

However there are cases where no instance of a service may be available in the availability zone of the client, in which case we want to fallback to another zone where the service has instances running.

What would you think about adding the ability to have fallback names in the rewrite plugin? That way if a lookup from a rewrite gives no results coredns moves to the next rewrite, until it runs out or one gives a successful result.

Basically it would be something like this (for a host in us-west-2a):
```
rewrite name regex ^(.*).domain$ us-west-2a.{1}.service.consul us-west-2b.{1}.service.consul us-west-2c.{1}.service.consul
```

**Comment by achille-roussel** [comment]
To be clear, this issue is not a feature request, we can submit a PR to implement the change, but we just want to make sure there are no fundamental issues that someone with more experience with coredns would know about.

**Comment by chrisohaver** [comment]
You might be able to (almost) do this with the `rewrite` and the `autopath` plugins.  However, your schema of the search domains would need to be altered from your example. 

If your naming scheme could instead be in the format:`{1}.us-west-2a.service.consul` ... `{1}.us-west-2b.service.consul` ... etc, then you could do 

```
rewrite name regex ^(.*).mydomain$ {1}.us-west-2a.service.consul
autopath availZones.conf
```  

`availZones.conf` would be a file containing something like ...
```
domain us-west-2a.service.consul us-west-2b.service.consul us-west-2c.service.consul
```

This would for example, rewrite `foo.mydomain` to `foo.us-west-2a.service.consul`, then the autopath plugin would detect `foo.us-west-2a.service.consul` as the first domain in the search path and try each domain in the list, in order, until it gets an answer. 

The hostname in the middle of the domain seems unusual to me.  But if the schema of the domains can't be altered, then you could extend autopath plugin parsing of the file to allow for inserting the left part of the query into the middle of the search domain... e.g. `us-west-2a.$.service.consul`.

[corrected availZones.conf file above]

**Comment by chrisohaver** [comment]
> ... if the schema of the domains can't be altered, then you could extend autopath plugin ...

Actually - accommodating the hostname in the middle of the domain would be more complex than I suggested.

**Comment by achille-roussel** [comment]
Thanks for the advise, I'll look into it. I'm not sure we have a lot of flexibility on the final name format since we have to comply with Consul's naming scheme (in our case the zone is a tag on the registered services).

**Comment by chrisohaver** [comment]
OK, accommodating the hostname in the middle of the domain would require handling it in two places, one in the substitution of hostname in each search domain in the path, and the other in the matching/extraction of  the hostname with the first search domain. 

For example (using `*` as a "wildcard" character):
 
- extraction:`zone1a.foo.service.consul` should match the search domain`zone1a.*.service.consul`, and `foo` extracted as the hostname
- substitution: hostname`foo` applied to search domain `zone1b.*.service.consul` should result in `zone1b.foo.service.consul` 

It would be a good idea to verify the rewrite + autopath concept works before diving in and making these mods.  I think it would, but I have not verified.

**Comment by fturib** [comment]
also, please consider that the REWRITE plugin is under transformation, adding a RESPONSE rewrite option see PR #1318

**Comment by chrisohaver** [comment]
> please consider that the REWRITE plugin is under transformation ...

Good to know. Though, I don’t think those changes will conflict with the proposed autopath solution.

**Comment by greenpau** [comment]
@achille-roussel, let me know whether I am getting at what you are trying to accomplish. Based on my reading, you want a conditional rewrite based on the availability of a record in a specific remote domain?

**Comment by achille-roussel** [comment]
I think what we're aiming for is something simpler that looks like this:
1. apply first rewrite rule
2. resolve the rewritten name
3. if no results were found, move to the second rewrite rule and try again

If I'm reading the code correctly, the current implementation stops at the first error it finds.

**Comment by greenpau** [comment]
@achille-roussel , I see. I have some thoughts about it.

It is not easy to accomplish what you want to do. It is an interesting use case :smile:

The rules are evaluated when you receive a request. Your evaluation of the existence of a record happens when you process a response (pls look at the code where we rewrite response).
On the failure to find a record, we need to re-initiate the request, while disabling the failed rewrite.

**Comment by johnbelamaric** [comment]
Yeah. I think this is more appropriately handled by `autopath` or something like it. It's essentially trying different names until you find a good one, which is what `autopath` does.

Basing on @chrisohaver suggestion, you may be able to create an option in `autopath` to act on a series of names as prefixes instead of suffixes. So maybe make `availZones.conf` look like:

```
prefix us-west-2a us-west-2b us-west-2c
```

instead of `domain`. Or you can put the path config directly in autopath config in the Corefile somehow.

**Comment by chrisohaver** [comment]
> you may be able to create an option in `autopath`

I was thinking of something along the lines of introducing an optional "anchor" in the autopath search path. e.g. using a `%` as the anchor ...

`domain   east.%.domain.org   west.%.domain.org`

The search path above would match a query of `east.foo.bar.domain.org` to `east.%.domain.org`, extracting `foo.bar`, and later (if `east.foo.bar.domain.org` doesn't have a record) insert `foo.bar` into `west.%.domain.org` resulting in `west.foo.bar.domain.org` ...

**Comment by chrisohaver** [comment]
> Or you can put the path config directly in autopath config in the Corefile

That would be a lot more convenient. Currently you have to point to a file.

The `prefix` option that @johnbelamaric suggests is slightly less flexible, but covers this case completely (since all search paths are in the same domain `service.consul`), it also would be easier to document, and a bit easier to implement, and it would require less string operations.

**Comment by miekg** [comment]
Note that Caddy supports `if` in multiple plugins: https://caddyserver.com/docs/rewrite#if

The conditions are all textual in Caddy, but we could add to that. Note that powerdns is experimenting with a LUA record - which sort of allows you to route/rewrite responses.

My preference would be using existing plugins (which may need a small tweak)

**Comment by achille-roussel** [comment]
Thanks a lot you all for your advises.

Here's what I came up with, and it seems to work:

- corefile.conf
```
service.consul:1053 {
    proxy . 127.0.0.1:1053
    rewrite name substring .service.consul .us-west-2a.service.consul
    autopath zones.conf
    errors
    log
}

us-west-2a.service.consul:1053 {
    proxy . 127.0.0.1:8600
    rewrite name regex (.*).us-west-2a.service.consul us-west-2a.{1}.service.consul
}

us-west-2b.service.consul:1053 {
    proxy . 127.0.0.1:8600
    rewrite name regex (.*).us-west-2b.service.consul us-west-2b.{1}.service.consul
}

us-west-2c.service.consul:1053 {
    proxy . 127.0.0.1:8600
    rewrite name regex (.*).us-west-2c.service.consul us-west-2c.{1}.service.consul
}
```

- zones.conf
```
search us-west-2a.service.consul us-west-2b.service.consul us-west-2c.service.consul
```

Then when making a request like this
```
$ dig @127.0.0.1 -p 1053 integrations.service.consul

; <<>> DiG 9.8.3-P1 <<>> @127.0.0.1 -p 1053 integrations.service.consul
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 17230
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;integrations.service.consul.	IN	A

;; Query time: 1 msec
;; SERVER: 127.0.0.1#1053(127.0.0.1)
;; WHEN: Fri Jan 26 09:16:09 2018
;; MSG SIZE  rcvd: 45
```
Consul receives those queries
```
    2018/01/26 09:19:22 [DEBUG] dns: request for {us-west-2a.integrations.service.consul. 1 1} (109.515µs) from client 127.0.0.1:59123 (udp)
    2018/01/26 09:19:22 [DEBUG] dns: request for {us-west-2b.integrations.service.consul. 1 1} (138.253µs) from client 127.0.0.1:55942 (udp)
    2018/01/26 09:19:22 [DEBUG] dns: request for {us-west-2c.integrations.service.consul. 1 1} (100.912µs) from client 127.0.0.1:57746 (udp)
```

To provide zone preference, all that needs to change is the first rewrite rule and the order of the zones in the autopath configuration file, which is trivial to generate at boot time.

**Comment by chrisohaver** [comment]
I didn't think of re-writing the autopath attempts too... brilliant!

**Comment by greenpau** [comment]
@achille-roussel , you hacked it! 👍 like the creativity! 😃

**Comment by greenpau** [comment]
@achille-roussel , as a side note, I would suggest re-writing the response too. See [here](https://github.com/coredns/coredns/tree/master/plugin/rewrite#response-rewrites)

e.g.:

```
us-west-2c.service.consul:1053 {
    proxy . 127.0.0.1:8600
    rewrite stop {
        name regex (.*).us-west-2c.service.consul us-west-2c.{1}.service.consul
        answer name us-west-2c\.(.*)\.service\.consul {1}.us-west-2c.service.consul
    }
}
```

**Comment by achille-roussel** [comment]
Cool, I'll add this 👍

**Comment by johnbelamaric** [comment]
@achille-roussel want to write a blog on this for coredns.io?

**Comment by achille-roussel** [comment]
Sure thing! Let me finish to deploy this to production to make sure this is all rock solid and we can get to this.

**Comment by achille-roussel** [comment]
@greenpau I tried to rewrite the answer like you suggested, it works for basic A queries but for some reasons dig doesn't like it on SRV queries:
```
$ dig @127.0.0.1 -p 1053 redis.us-west-2b.service.consul

; <<>> DiG 9.8.3-P1 <<>> @127.0.0.1 -p 1053 redis.us-west-2b.service.consul
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27246
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;redis.us-west-2b.service.consul. IN	A

;; ANSWER SECTION:
redis.us-west-2b.service.consul. 0 IN	A	127.0.0.1

;; Query time: 2 msec
;; SERVER: 127.0.0.1#1053(127.0.0.1)
;; WHEN: Sat Jan 27 09:28:00 2018
;; MSG SIZE  rcvd: 65
```
```
$ dig @127.0.0.1 -p 1053 redis.us-west-2b.service.consul SRV
;; Question section mismatch: got us-west-2b.redis.service.consul/SRV/IN
```
With just the `rewrite name` rule here's what it gives:
```
$ dig @127.0.0.1 -p 1053 redis.service.consul SRV

; <<>> DiG 9.8.3-P1 <<>> @127.0.0.1 -p 1053 redis.service.consul SRV
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 819
;; flags: qr aa rd; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;redis.service.consul.		IN	SRV

;; ANSWER SECTION:
redis.us-west-2a.service.consul. 0 IN	CNAME	us-west-2b.redis.service.consul.
us-west-2b.redis.service.consul. 0 IN	SRV	1 1 6379 453fdf18f226.node.dc1.consul.

;; ADDITIONAL SECTION:
453fdf18f226.node.dc1.consul. 0	IN	A	127.0.0.1

;; Query time: 11 msec
;; SERVER: 127.0.0.1#1053(127.0.0.1)
;; WHEN: Sat Jan 27 09:29:02 2018
;; MSG SIZE  rcvd: 144
```
I wonder if maybe we can modify the rewrite plugin to always inject a CNAME for the original name in the answer to trick resolvers that would otherwise refuse the query? I have to admit this is beyond my level of knowledge of the DNS protocol so I'm not sure what side effects it could have. For the use case I'm tackling I think the current behavior is OK since it's all internal service name resolutions and we have full control over the code and infrastructure.

**Comment by greenpau** [comment]
> I tried to rewrite the answer like you suggested, it works for basic A queries but for some reasons dig doesn't like it on SRV queries

@achille-roussel , I would need to look at it.

> I have to admit this is beyond my level of knowledge of the DNS protocol so I'm not sure what side effects it could have.

@achille-roussel , I am in the same ⛵️ 😄 By and large, we are all learning! 😃

**Comment by achille-roussel** [comment]
@greenpau sure, I should have included it,
- zonefile.conf:
```
search us-west-2a.service.consul us-west-2b.service.consul us-west-2c.service.consul
```
- corefile.conf (working)
```
service.consul {
    proxy . 127.0.0.1
    rewrite name substring .service.consul .us-west-2a.service.consul
    autopath /zonefile.conf
}
us-west-2a.service.consul {
    proxy . 127.0.0.1:8600
    rewrite name regex (.*)\.us-west-2a\.service\.consul us-west-2a.{1}.service.consul
}
us-west-2b.service.consul {
    proxy . 127.0.0.1:8600
    rewrite name regex (.*)\.us-west-2b\.service\.consul us-west-2b.{1}.service.consul
}
us-west-2c.service.consul {
    proxy . 127.0.0.1:8600
    rewrite name regex (.*)\.us-west-2c\.service\.consul us-west-2c.{1}.service.consul
}
```
- corefile.conf (not working)
```
service.consul {
    proxy . 127.0.0.1
    rewrite stop {
        name regex (.*)\.service.consul {1}.us-west-2a.service.consul
        answer name (.*)\.(us-west-2a|us-west-2b|us-west-2c)\.service\.consul {1}.service.consul
    }
    autopath /zonefile.conf
}
us-west-2a.service.consul {
    proxy . 127.0.0.1:8600
    rewrite stop {
        name regex (.*)\.us-west-2a\.service\.consul us-west-2a.{1}.service.consul
        answer name us-west-2a\.(.*)\.service\.consul {1}.us-west-2a.service.consul
    }
}
us-west-2b.service.consul {
    proxy . 127.0.0.1:8600
    rewrite stop {
        name regex (.*)\.us-west-2b\.service\.consul us-west-2b.{1}.service.consul
        answer name us-west-2b\.(.*)\.service\.consul {1}.us-west-2b.service.consul
    }
}
us-west-2c.service.consul {
    proxy . 127.0.0.1:8600
    rewrite stop {
        name regex (.*)\.us-west-2c\.service\.consul us-west-2c.{1}.service.consul
        answer name us-west-2c\.(.*)\.service\.consul {1}.us-west-2c.service.consul
    }
}
```
I have a `redis` service registered to consul, tagged with `us-west-2b`. With the answer rewrite only A queries to `redis.us-west-2b.service.consul` are accepted by dig, querying `redis.service.consul` or `redis-us-west-2b.service.consul` as a SRV query yields the kind of errors that I showed above.
I verified that the regex patterns in the answer rewrite match outside of the context of coredns (because I initially thought that I had made a mistake there), so I'm not sure what else it may be.

**Comment by greenpau** [comment]
@achille-roussel , the https://github.com/coredns/coredns/commit/f636930c5cb72537e34856f2eb0e68ec17202f0e was merged. Could you please resync and test your SRV use-case?

**Comment by achille-roussel** [comment]
I just checked and it seems to be working well, I tried A, and SRV with a _classic_ name format as well as a RFC 2782 format and got a response every time.

Thanks for the fix 👍
